### PR TITLE
Fix MP Config FAT tests so they can run against mpConfig 2.0

### PR DESCRIPTION
**Pull request by Joseph-Cass**
* Created mpConfig2.0 specific tests within the 2.0 FAT bucket where the functionality has changed from 1.X-> 2.0
* Created mpConfig1.x specific tests/packages within the 1.x FAT buckets, where the Liberty impl has extra or different functionality from SmallRye impl

The code depends on recent changes made to SmallRye Config which may not yet have been uploaded to DHE/Maven Central- so some test failures may occur.

### CODE REVIEW

File: dev/com.ibm.ws.microprofile.config.1.1_fat/fat/src/com/ibm/ws/microprofile/config/fat/suite/SharedShrinkWrapApps.java [hunk 1]
```diff
@@ -26,7 +26,7 @@
     public static JavaArchive getTestAppUtilsJar() {
         final String LIB_NAME = "testAppUtils";
         JavaArchive testAppUtils = ShrinkWrap.create(JavaArchive.class, LIB_NAME + ".jar")
-                        .addPackage("com.ibm.ws.microprofile.appConfig.test.utils");
+                                             .addPackage("com.ibm.ws.microprofile.appConfig.test.utils");
```
Comment (Emily-Jiang) [inline dev/com.ibm.ws.microprofile.config.1.1_fat/fat/src/com/ibm/ws/microprofile/config/fat/suite/SharedShrinkWrapApps.java:29]: you got some tabs here maybe

File: dev/com.ibm.ws.microprofile.config.1.1_fat/fat/src/com/ibm/ws/microprofile/config/fat/suite/SharedShrinkWrapApps.java [hunk 2]
```diff
@@ -38,22 +38,24 @@ public static WebArchive cdiConfigServerApps() {
         }
 
         cdiConfig_war = ShrinkWrap.create(WebArchive.class, APP_NAME + ".war")
-                        .addAsManifestResource(new File("test-applications/" + APP_NAME + ".war/resources/META-INF/permissions.xml"), "permissions.xml")
-                        .addAsManifestResource(new File("test-applications/" + APP_NAME + ".war/resources/META-INF/services/org.eclipse.microprofile.config.spi.ConfigSource"),
-                                               "services/org.eclipse.microprofile.config.spi.ConfigSource")
-                        .addAsManifestResource(new File("test-applications/" + APP_NAME + ".war/resources/META-INF/services/org.eclipse.microprofile.config.spi.Converter"),
-                                               "services/org.eclipse.microprofile.config.spi.Converter")
-                        .addAsLibrary(cdiConfigJar());
+                                  .addAsManifestResource(new File("test-applications/" + APP_NAME + ".war/resources/META-INF/permissions.xml"), "permissions.xml")
```
Comment (Emily-Jiang) [inline dev/com.ibm.ws.microprofile.config.1.1_fat/fat/src/com/ibm/ws/microprofile/config/fat/suite/SharedShrinkWrapApps.java:41]: can you fix the format?

File: dev/com.ibm.ws.microprofile.config.1.1_fat/fat/src/com/ibm/ws/microprofile/config/fat/tests/LibertySpecificConfigTests.java [hunk 3]
```diff
@@ -0,0 +1,111 @@
...
+import componenttest.rules.repeater.RepeatTests;
+import componenttest.topology.impl.LibertyServer;
+import componenttest.topology.utils.FATServletClient;
+
+/**
+ * These tests test Config functionality specific to Liberty's implementation (mpConfig 1.x) of MicroProfile Config specification.
+ *
+ * For mpConfig versions <2.0, Open Liberty implemented some extra pieces of functionality beyond the MicroProfile Config specification.
+ *
+ * In additon, some tests check for error messages defined by Liberty, and others check functionality which changed between mpConfig 1.4 -> 2.0.
```
Comment (Emily-Jiang) [inline dev/com.ibm.ws.microprofile.config.1.1_fat/fat/src/com/ibm/ws/microprofile/config/fat/tests/LibertySpecificConfigTests.java:48]: Please list what these tests for.

File: dev/com.ibm.ws.microprofile.config.1.1_fat/test-applications/brokenCDIConfig.war/src/com/ibm/ws/microprofile/appConfig/cdi/broken/test/TypeWithValidConverter.java [hunk 4]
```diff
@@ -1,5 +1,5 @@
 /*******************************************************************************
- * Copyright (c) 2019 IBM Corporation and others.
+ * Copyright (c) 2018 IBM Corporation and others.
```
Comment (Emily-Jiang) [inline dev/com.ibm.ws.microprofile.config.1.1_fat/test-applications/brokenCDIConfig.war/src/com/ibm/ws/microprofile/appConfig/cdi/broken/test/TypeWithValidConverter.java:2]: why did you change the year?

File: dev/com.ibm.ws.microprofile.config.1.1_fat/test-applications/brokenCDIConfig.war/src/com/ibm/ws/microprofile/appConfig/cdi/broken/test/TypeWithValidConverter.java [hunk 5]
```diff
@@ -8,23 +8,21 @@
...
 
 /**
  *
  */
-@WebServlet("/")
-public class BadObserverServlet extends FATServlet {
+public class TypeWithValidConverter {
 
-    @Inject
-    BadObserverBean bean;
```
Comment (Emily-Jiang) [inline dev/com.ibm.ws.microprofile.config.1.1_fat/test-applications/brokenCDIConfig.war/src/com/ibm/ws/microprofile/appConfig/cdi/broken/test/TypeWithValidConverter.java:25]: Why did you delete this?

**Review by Emily-Jiang — COMMENTED** [review]
Can you document what tests are for liberty specific tests and list the behaviour etc?

**Comment by Joseph-Cass** [comment]
#build

File: dev/com.ibm.ws.microprofile.config.1.4_fat/fat/src/com/ibm/ws/microprofile/config14/test/LibertySpecificConfigTests.java [hunk 6]
```diff
@@ -1,5 +1,5 @@
 /*******************************************************************************
```
Comment (Joseph-Cass) [inline dev/com.ibm.ws.microprofile.config.1.4_fat/fat/src/com/ibm/ws/microprofile/config14/test/LibertySpecificConfigTests.java:1]: Note to reviewer(s): Git has muddled up a file creation + deletion for a file rename here.

File: dev/com.ibm.ws.microprofile.config.1.1_fat/test-applications/brokenCDIConfig.war/src/com/ibm/ws/microprofile/appConfig/cdi/broken/test/TypeWithNoConverter.java [hunk 7]
```diff
@@ -1,5 +1,5 @@
 /*******************************************************************************
```
Comment (Joseph-Cass) [inline dev/com.ibm.ws.microprofile.config.1.1_fat/test-applications/brokenCDIConfig.war/src/com/ibm/ws/microprofile/appConfig/cdi/broken/test/TypeWithNoConverter.java:1]: Note to reviewer(s): Git has muddled up a file creation + deletion for a file rename here.

File: dev/com.ibm.ws.microprofile.config.1.1_fat/test-applications/brokenCDIConfig.war/src/com/ibm/ws/microprofile/appConfig/cdi/broken/test/TypeWithValidConverter.java [hunk 8]
```diff
@@ -1,5 +1,5 @@
 /*******************************************************************************
```
Comment (Joseph-Cass) [inline dev/com.ibm.ws.microprofile.config.1.1_fat/test-applications/brokenCDIConfig.war/src/com/ibm/ws/microprofile/appConfig/cdi/broken/test/TypeWithValidConverter.java:1]: Note to reviewer(s): Git has muddled up a file creation + deletion for a file rename here.

File: dev/com.ibm.ws.microprofile.config.1.1_fat/test-applications/cdiConfig.jar/src/com/ibm/ws/microprofile/appConfig/cdi/libertyTests/LibertyBuiltInConverterTestServlet.java [hunk 9]
```diff
@@ -1,5 +1,5 @@
 /*******************************************************************************
```
Comment (Joseph-Cass) [inline dev/com.ibm.ws.microprofile.config.1.1_fat/test-applications/cdiConfig.jar/src/com/ibm/ws/microprofile/appConfig/cdi/libertyTests/LibertyBuiltInConverterTestServlet.java:1]: Note to reviewer(s): Git has muddled up a file creation + deletion for a file rename here.

File: dev/io.openliberty.microprofile.config.2.0.internal_fat/fat/src/io/openliberty/microprofile/config/internal_fat/apps/brokenInjection/MissingConfigPropertyBean.java [hunk 10]
```diff
@@ -1,5 +1,5 @@
 /*******************************************************************************
```
Comment (Joseph-Cass) [inline dev/io.openliberty.microprofile.config.2.0.internal_fat/fat/src/io/openliberty/microprofile/config/internal_fat/apps/brokenInjection/MissingConfigPropertyBean.java:1]: Note to reviewer(s): Git has muddled up a file creation + deletion for a file rename here.

**Comment by Joseph-Cass** [comment]
The 1.3 bucket was missed...

#build
#spawn.fullfat.buckets= com.ibm.ws.microprofile.config.1.3_fat

File: dev/com.ibm.ws.microprofile.config.1.1_fat/fat/src/com/ibm/ws/microprofile/config/fat/tests/LibertySpecificConfigTests.java (hunk 3 above)
Comment (Joseph-Cass) [reply dev/com.ibm.ws.microprofile.config.1.1_fat/fat/src/com/ibm/ws/microprofile/config/fat/tests/LibertySpecificConfigTests.java]: Any tests within `LibertySpecificConfigTests.java` classes are tests for functionality specific to Liberty's implementation (mpConfig < 2.0) of MP Config (as explained in the comment).

The tests should be self-explanatory, or have comments explaining their purpose.

File: dev/io.openliberty.microprofile.config.2.0.internal_fat/fat/src/io/openliberty/microprofile/config/internal_fat/apps/badobserver/TestObserver.java [hunk 11]
```diff
@@ -8,7 +8,7 @@
  * Contributors:
  *     IBM Corporation - initial API and implementation
  *******************************************************************************/
-package io.openliberty.microprofile.config.internal_fat.apps.badobserver;
+package io.openliberty.microprofile.config.internal_fat.apps.badObserver;
```
Comment (Azquelt) [inline dev/io.openliberty.microprofile.config.2.0.internal_fat/fat/src/io/openliberty/microprofile/config/internal_fat/apps/badobserver/TestObserver.java:11]: Package name has a capitalized `O` whereas the corresponding directory does not.

File: dev/com.ibm.ws.microprofile.config.1.1_fat/test-applications/cdiConfig.jar/src/com/ibm/ws/microprofile/appConfig/cdi/beans/ConfigFieldInjectionBean.java [hunk 12]
```diff
@@ -203,25 +167,32 @@ public URL getOPTIONAL_MISSING_KEY() {
...
+
+    @Override
+    public Pizza getPIZZA_KEY() {
+        return null;
+    }
+
+    @Override
+    public Pizza getPIZZA_MISSING_KEY() {
+        return null;
+    }
```
Comment (Azquelt) [inline dev/com.ibm.ws.microprofile.config.1.1_fat/test-applications/cdiConfig.jar/src/com/ibm/ws/microprofile/appConfig/cdi/beans/ConfigFieldInjectionBean.java:197]: These methods which return `null` look like they're not referenced anywhere any more?

File: dev/com.ibm.ws.microprofile.config.1.1_fat/test-applications/cdiConfig.jar/src/com/ibm/ws/microprofile/appConfig/cdi/beans/ConfigFieldInjectionInterface.java [hunk 13]
```diff
@@ -48,6 +48,8 @@
 
     public URL getOPTIONAL_MISSING_KEY();
 
+    public String getMISSING_KEY();
+
     public Pizza getPIZZA_KEY();
```
Comment (Azquelt) [inline dev/com.ibm.ws.microprofile.config.1.1_fat/test-applications/cdiConfig.jar/src/com/ibm/ws/microprofile/appConfig/cdi/beans/ConfigFieldInjectionInterface.java:53]: Not sure that this interface is actually referenced anywhere apart from the bean that implements it. Maybe it could be taken out?

Comment (Joseph-Cass) [reply dev/com.ibm.ws.microprofile.config.1.1_fat/test-applications/cdiConfig.jar/src/com/ibm/ws/microprofile/appConfig/cdi/beans/ConfigFieldInjectionInterface.java:53]: Yes, you're right, only ConfigFieldInjectionBean.java implements this interface. I'd be happy to take it out. This would also resolve this comment: https://github.com/OpenLiberty/open-liberty/pull/14375#discussion_r504081213

**Comment by Joseph-Cass** [comment]
Build failed due to something nothing to do with these changes. Trying again:

#build
#spawn.fullfat.buckets=com.ibm.ws.microprofile.config.1.3_fat

**Comment by Joseph-Cass** [comment]
Rebased to include https://github.com/OpenLiberty/open-liberty/pull/14584. All FAT tests should pass now.

#build
#spawn.fullfat.buckets=com.ibm.ws.microprofile.config.1.3_fat

**Comment by Joseph-Cass** [comment]
#build
#spawn.fullfat.buckets=com.ibm.ws.microprofile.config.1.3_fat

**Comment by Joseph-Cass** [comment]
Lists of FATs fixed here: https://github.com/OpenLiberty/open-liberty/issues/13644#issuecomment-686569355

File: dev/io.openliberty.microprofile.config.2.0.internal_fat/fat/src/io/openliberty/microprofile/config/internal_fat/apps/defaultSources/DefaultSourcesTestServlet.java [hunk 14]
```diff
@@ -0,0 +1,74 @@
...
+
+            if (!systemValue.getValue().isEmpty()) {
+
+                String expectedValue = systemValue.getValue();
+                String actualValue = config.getValue(expectedKey, expectedKey.getClass());
+
+                Assert.assertTrue("The system value for " + expectedKey + " should be " + expectedValue + ", not " + actualValue,
+                                  expectedValue.equals(actualValue));
+            }
+        }
```
Comment (Azquelt) [inline dev/io.openliberty.microprofile.config.2.0.internal_fat/fat/src/io/openliberty/microprofile/config/internal_fat/apps/defaultSources/DefaultSourcesTestServlet.java:72]: This test isn't correct any more if any of the system properties or environment variables contain `${foo}` as this would trigger property expansion so the input value won't match the value retrieved from config.

This does seem to happen on some linux systems.

**Review by Azquelt — APPROVED** [review]

**Comment by Joseph-Cass** [comment]
#build
#spawn.fullfat.buckets=com.ibm.ws.microprofile.config.1.3_fat

**Comment by Joseph-Cass** [comment]
#run-libby-bot
